File: Spring_2019/Questions/Medium/scoreOfParenthesis/scoreOfParenthesis.go
package score

import (
	"math"
	"strings"
)

func scoreOfParenthesesUsingStack(parens string) float64 {
	stack := []float64{0}

	for _, val := range parens {
		if strings.Compare(string(val), "(") == 0 {
			stack = append(stack, 0)
		} else {
			x := stack[len(stack)-1]
			stack = stack[0 : len(stack)-1]
			stack[len(stack)-1] += math.Max(float64(2*x), float64(1))
		}
	}

	return stack[0]
}

func scoreOfParenthesesUsingCores(parens string) float64 {
	ans := float64(0)
	bal := float64(0)

	for index, val := range parens {
		if strings.Compare(string(val), "(") == 0 {
			bal++
		} else {
			bal--
			if strings.Compare(string(parens[index-1]), "(") == 0 {
				add := math.Pow(float64(2), bal)
				ans += add
			}
		}
	}

	return ans
}

File: Spring_2019/March_5/scoreOfParenthesis/scoreOfParenthesis.py
import unittest

class Solution(unittest.TestCase):
    @staticmethod
    def scoreOfParenthesesUsingStack(S):
        stack = [0] #The score of the current frame

        for x in S:
            if x == '(':
                stack.append(0)
            else:
                v = stack.pop()
                stack[-1] += max(2 * v, 1)

        return stack.pop()

    @staticmethod
    def scoreOfParenthesesUsingCores(S):
        ans = bal = 0
        for i, x in enumerate(S):
            if x == '(':
                bal += 1
            else:
                bal -= 1
                if S[i-1] == '(':
                    ans += 1 << bal
        return ans

    def testStack(self):
        testCase = '(()(()))'
        ans = 6
        self.assertEqual(self.scoreOfParenthesesUsingStack(testCase), ans)

    def testCountingCores(self):
        testCase = '(()(()))'
        ans = 6
        self.assertEqual(self.scoreOfParenthesesUsingCores(testCase), ans)

if __name__ == '__main__':
    unittest.main()
